### ISSUE TITLE
Delay in edges does not work

### ISSUE THREAD

**Issue by jajcayn** [pyrates-neuroscience/PyRates#9]
Defining a delay in an edge yield an error of shape mismatch.

My implementation:
```yaml
- [PYR/PyramidalCPO/V_out, TCR/ThalamicPRO/V_in, *LE, {*c : 1., weight: 1., delay: 0.001}]
```
hence I want to introduce a delay of 1ms (all my temporal variables are defined in seconds, hence delay=1e-3).
Both backends yield the following error:
```
ValueError: Invalid index shape. Operation has shape (12,), but received an index of length 2
```
using dt=1e-4. When using dt=1e-5 the shape mismatch becomes (102,) so I guess the discretization of delays or setting the correct length of history are the possible culprits.

I could try to fix it myself, if you'd point me in the right direction - the backend codebase is pretty large and on the first look, I couldn't see the associated methods.

Cheers,
N.

P.S. if I am not mistaken, also the associated test (`tests/test_compute_graph.py`) fails, I tested on MacOS python3.7 and Ubuntu18, python3.6

**Comment by dafrose** [comment]
Hi Nikola,

thank you very much for your feedback. Which version are you using? From current github or latest release on PyPI?

Could you please provide the full template for testing? Github allows uploading of .txt files, so if .yml does not work, you could try that. 

Best,
Daniel

**Comment by jajcayn** [comment]
So, after some additional testing - the current PyPI version 0.8.0 throws the above-written error, however, when I install from GitHub repo the current master, it works!

**Comment by jajcayn** [comment]
Final comment:
adding output, when I added printing statements to `_cache_op_graph` wrapper:
```
869794232570078598
CorticalPRO <OperatorIR(('r = r_max / (1. + exp(sigma*(V_thr - V_in)))',)), hash = -6391445111721950068 >
PyramidalRCO <OperatorIR(('d/dt * I = I_t', 'd/dt * I_t =  G * omega * r - omega^2 * I - 2. * omega * I_t')), hash = 1186711113592817207 >
PyramidalCPO <OperatorIR(('V_out = k * I + I_modulatory + I_driver',)), hash = 8410447827851721302 >

1626518669850294754
CorticalPRO <OperatorIR(('r = r_max / (1. + exp(sigma*(V_thr - V_in)))',)), hash = -6391445111721950068 >
ExcitatoryRCO <OperatorIR(('d/dt * I = I_t', 'd/dt * I_t =  G * omega * r - omega^2 * I - 2. * omega * I_t')), hash = 1186711113592817207 >
ExcitatoryCPO <OperatorIR(('V_out = k * I + I_modulatory + I_driver',)), hash = -5857010408256079830 >

1626518669850294754
CorticalPRO <OperatorIR(('r = r_max / (1. + exp(sigma*(V_thr - V_in)))',)), hash = -6391445111721950068 >
InhibitorySlowRCO <OperatorIR(('d/dt * I = I_t', 'd/dt * I_t =  G * omega * r - omega^2 * I - 2. * omega * I_t')), hash = 1186711113592817207 >
InhibitorySlowCPO <OperatorIR(('V_out = k * I + I_modulatory + I_driver',)), hash = -5857010408256079830 >

1626518669850294754
CorticalPRO <OperatorIR(('r = r_max / (1. + exp(sigma*(V_thr - V_in)))',)), hash = -6391445111721950068 >
InhibitoryFastRCO <OperatorIR(('d/dt * I = I_t', 'd/dt * I_t =  G * omega * r - omega^2 * I - 2. * omega * I_t')), hash = 1186711113592817207 >
InhibitoryFastCPO <OperatorIR(('V_out = k * I + I_modulatory + I_driver',)), hash = -5857010408256079830 >
```

the first line is hash of a node, followed by a dictionary of operators. The first node (`PYR`) has some hash value and their operators, the second node (`EXC`) the same, but the following two nodes have the same hash value as `EXC` even they are defined using different operators. I guess that is why the nodes then exhibit wrong addresses as described above.

**Comment by dafrose** [comment]
Hey Nikola,
thanks a lot for your valuable feedback. I just wanted to let you know that we are working on it. Dev time turned out to be scarce during parental leave, but I'll be back in office soon and hope to finally get these issues done. 

Best,
Daniel

**Comment by Richert** [comment]
Closing this now. The bug with the edge delays has been solved with 0.9.0. Discretization of delayed-differential equations was stabilized and, additionally, the possibility was added to work with gamma-distributed delays instead, by using a  `spread` keyword:

```yaml
- [PYR/PyramidalCPO/V_out, TCR/ThalamicPRO/V_in, *LE, {*c : 1., weight: 1., delay: 0.001, spread: 0.0005}]
```

This way, `delay` and `spread` resemble the mean and variance of a gamma delay kernel function that the source signal is convoluted with. This is a way to prevent discretization inaccuracies that may arise when working with scalar delays.
